File: frontend/src/app/core/services/attendance.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { environment } from '../../../environments/environment';
import { Attendance, Page } from '../models/api.models';

@Injectable({ providedIn: 'root' })
export class AttendanceService {
  private http = inject(HttpClient);
  private url = `${environment.apiUrl}/attendance`;
  record(body: { studentId: number; courseId: number; attendancePercentage: number }) { return this.http.post<Attendance>(this.url, body); }
  update(id: number, body: { studentId: number; courseId: number; attendancePercentage: number }) { return this.http.put<Attendance>(`${this.url}/${id}`, body); }
  delete(id: number) { return this.http.delete<void>(`${this.url}/${id}`); }
  byStudent(studentId: number) { return this.http.get<Attendance[]>(`${this.url}/student/${studentId}`); }
  defaulters() { return this.http.get<Page<Attendance>>(`${this.url}/defaulters?size=50`); }
}

File: backend/src/main/java/com/portal/studentportal/service/AttendanceService.java
package com.portal.studentportal.service;

import com.portal.studentportal.dto.attendance.AttendanceRequest;
import com.portal.studentportal.dto.attendance.AttendanceResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface AttendanceService {
    AttendanceResponse record(AttendanceRequest request);
    AttendanceResponse update(Long id, AttendanceRequest request);
    void delete(Long id);
    List<AttendanceResponse> byStudent(Long studentId);
    Page<AttendanceResponse> defaulters(Pageable pageable);
}
